src/force_latest_compatible_version.jl

```diff
@@ -0,0 +1,75 @@
+    direct_deps = load_direct_deps(ctx.env)
+    direct_deps_uuids = getproperty.(direct_deps, Ref(:uuid))
+    nonstdlib_direct_deps_uuids = filter(x -> x ∉ keys(stdlibs()), direct_deps_uuids)
+    return _force_latest_compatible_version(ctx, nonstdlib_direct_deps_uuids; target_name)
+end
+
+function _force_latest_compatible_version(ctx::Types.Context, uuid_list::AbstractVector{Base.UUID};
+                                          target_name = nothing)
+    isempty(uuid_list) && return true
+    results = _force_latest_compatible_version.(Ref(ctx), uuid_list; target_name)
+    return all(results)
+end
+
+function _warn_no_compat(has_compat::Bool,
+                         target_name::AbstractString,
+                         name::AbstractString,
+                         uuid::Base.UUID,
+                         active_version::Base.VersionNumber)
+    if !has_compat
+        if name != target_name
+            @warn(
+                "Package does not have a [compat] entry",
+                name, uuid, active_version, target_name,
+            )
+        end
+    end
+    return nothing
+end
+
+function _force_latest_compatible_version(ctx::Types.Context, uuid::Base.UUID;
+                                          target_name= nothing)
+    manifest_deps = load_manifest_deps(ctx.env.manifest)
+    manifest_deps_uuids = getproperty.(manifest_deps, Ref(:uuid))
+    idx = only(findall(manifest_deps_uuids .== uuid))
+    dep = manifest_deps[idx]
+    name = dep.name
+    active_version = dep.version
+    has_compat = haskey(ctx.env.project.compat, name)
+    _warn_no_compat(has_compat, target_name, name, uuid, active_version)
+    has_compat || return true
+    compat_entry_string = ctx.env.project.compat[name]
+    latest_compatible_version = _get_latest_compatible_version(ctx, uuid, compat_entry_string)
+    result = active_version >= latest_compatible_version
+    if !result
+        @error(
+            "Package is not at the latest compatible version",
+            name, uuid, active_version,
+            latest_compatible_version, compat_entry_string,
+        )
+    end
+    return result
+end
+
+function _get_latest_compatible_version(ctx::Types.Context, uuid::Base.UUID,
+                                        compat_entry_string::AbstractString)
+    all_registered_versions = _get_all_registered_versions(ctx, uuid)
+    compat_spec = Pkg.Types.semver_spec(compat_entry_string)
+    compatible_versions = filter(x -> x in compat_spec, all_registered_versions)
+    latest_compatible_version = maximum(compatible_versions)
+    return latest_compatible_version
+end
+
+function _get_all_registered_versions(ctx::Types.Context, uuid::Base.UUID)
+    versions = Base.VersionNumber[]
+    for reg in ctx.registries
+        pkg = get(reg, uuid, nothing)
+        pkg == nothing && continue
+        info = Registry.registry_info(pkg)
+        append!(versions, keys(info.version_info))
+    end
+    unique!(versions)
+    sort!(versions)
```

Comment:
Is there a need to `sort` here. You could use a `Set{VersionNumber}` if you want to only gather the unique versions.

Reply:
Done.